text/3446-store.md

```diff
@@ -0,0 +1,998 @@
+        _handle: Self::Handle,
+        old_layout: Layout,
+        new_layout: Layout,
+    ) -> Result<(Self::Handle, usize), AllocError> {
+        debug_assert!(
+            new_layout.size() >= old_layout.size(),
+            "{new_layout:?} must have a greater size than {old_layout:?}"
+        );
+
+        Self::validate_layout(new_layout)?;
+
+        let pointer = self.0.get() as *mut u8;
+
+        //  Safety:
+        //  -   Both starting and resulting pointers are in bounds of the same allocated objects as `old_layout` fits
+        //      `pointer`, as per the pre-conditions of `grow_zeroed`.
+        //  -   The offset does not overflow `isize` as `old_layout.size()` does not.
+        let pointer = unsafe { pointer.add(old_layout.size()) };
+
+        //  Safety:
+        //  -   `pointer` is valid, since `self` is valid.
+        //  -   `pointer` points to at an area of at least `mem::size_of::<T>() - old_layout.size()`.
+        //  -   Access to the next `mem::size_of::<T>() - old_layout.size()` bytes is exclusive.
+        unsafe { ptr::write_bytes(pointer, 0, mem::size_of::<T>() - old_layout.size()) };
+
+        Ok(((), mem::size_of::<T>()))
+    }
+}
+
+//  Safety:
+//  -   `self.resolve(handle)` always returns the same address, as long as `self` doesn't move.
+unsafe impl<T> StableStore for InlineSingleStore<T> {}
+
+impl<T> fmt::Debug for InlineSingleStore<T> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> Result<(), fmt::Error> {
+        let layout = Layout::new::<T>();
+
+        f.debug_struct("InlineSingleStore")
+            .field("size", &layout.size())
+            .field("align", &layout.align())
+            .finish()
+    }
+}
+
+impl<T> InlineSingleStore<T> {
+    fn validate_layout(layout: Layout) -> Result<(), AllocError> {
+        let own = Layout::new::<T>();
+
+        if layout.align() <= own.align() && layout.size() <= own.size() {
+            Ok(())
+        } else {
+            Err(AllocError)
+        }
+    }
+}
+```
+
+And that's it!
+
+I need not implement `MultipleStore`, and thus do not have to track allocations and deallocations. And I need not
+implement `PinningStore`, and thus do not have to ensure memory address stability across moves.
+
+More examples of `Store` can be found at https://github.com/matthieu-m/storage/tree/main/src/store, including an inline
+bump store.
+
+
+#   Reference-level explanation
+
+##  Overview
+
+This RFC introduces 4 new traits.
+
+The core of this RFC is the `Store` trait:
+
+```rust
+/// Allocates and deallocates handles to blocks of memory, which can be temporarily resolved to pointers to actually
+/// access said memory.
+pub unsafe trait Store {
+    /// Handle to a block of memory.
+    type Handle: Copy;
+
+    /// Returns a dangling handle, always invalid.
+    fn dangling() -> Self::Handle;
+
+    /// Return a pointer to the block of memory associated to `handle`.
+    unsafe fn resolve(&self, handle: Self::Handle) -> NonNull<u8>;
```

Comment:
To be clear, you are talking _specifically_ about a `SmallSingleStore`, that is, a `Store` which can store a _single_ allocation at once, as a small store storing multiple allocations could not possibly determine whether the storage is in-line or off-line based on the size of a single allocation.

It's a bit tough to decide here, as both thin pointers and `SmallSingleStore` seem like fairly niche usecases.

There are other options. For example, the `resolve` method could be split into a separate trait, something along the lines of:

```rust
trait StoreResolver: Store {
    fn resolve(&self, handle: Self::Handle) -> NonNull<u8>;
}

trait StoreLayoutResolver: Store {
    fn resolve_with_layout(&self, handle: Self::Handle, layout: Layout) -> NonNull<u8>;
}
```

Users could then pick resolution they wish for, and an optimized `SmallSingleStore` would only implement the latter and be incompatible with `ThinBox`.

I'll mention it in alternatives.

Reply:
I typically haven't specified `XxSingleStorage`, because storages only store one thing by default, and requesting multiple simultaneous handles from a storage is the uncommon case.

Single storages are the more interesting and common case because of RAII concerns. Even for multi storages, they're primarily consumed by multiple handles which are each used to store a single thing.

Perhaps this bias comes from only really having implemented collections which either refcounted[^1] each allocation or tries very hard to get everything into a single continuous allocation, rather than larger singly-owned collections which manage multiple allocated regions.

[^1]: ... Interesting thought; perhaps what `Arc<T, A>` wants isn't actually a sharing storage and to duplicate the allocator handle over each and every `Arc`, but a move-stable storage and to put the allocator handle in the heap allocation so the one allocator handle is available to whichever `Arc` ultimately frees the allocation 🤔